core/divelist.c

```diff
@@ -844,41 +846,39 @@ struct dive *last_selected_dive()
 	return ret;
 }
 
+static void unregister_dive_from_table(struct dive_table *table, int idx)
+{
+	int i;
+	for (i = idx; i < table->nr - 1; i++)
+		table->dives[i] = table->dives[i + 1];
+	table->dives[--table->nr] = NULL;
+}
+
 /* remove a dive from the trip it's associated to, but don't delete the
  * trip if this was the last dive in the trip. the caller is responsible
- * for removing the trip, if the trip->nrdives went to 0.
+ * for removing the trip, if the trip->dives.nr went to 0.
  */
 struct dive_trip *unregister_dive_from_trip(struct dive *dive, short was_autogen)
 {
-	struct dive *next, **pprev;
 	dive_trip_t *trip = dive->divetrip;
+	int idx;
 
 	if (!trip)
 		return NULL;
 
-	/* Remove the dive from the trip's list of dives */
-	next = dive->next;
-	pprev = dive->pprev;
-	*pprev = next;
-	if (next)
-		next->pprev = pprev;
-
+	idx = get_idx_in_table(&trip->dives, dive);
+	if (idx)
+		unregister_dive_from_table(&trip->dives, idx);
 	dive->divetrip = NULL;
-	if (was_autogen)
```

Comment:
I notice that you no longer handle the `was_autogen` flag (which has become unused).
That seems like it would be a problem.

Reply:
Yes, I will have to recheck that change.